Add contains_only matcher

danodonovan opened the pull request.

mwilliamson commented on README.rst:
```diff
@@ -97,6 +97,14 @@ For instance, ``has_attrs(name="bob")`` is equivalent to ``has_attrs(name=equal_
       # Matches ["a", "b"], ["b", "a"] and ["a", "c", "b"]
       # but not ["a", "c"] nor ["a"]
 
+* ``contains_only(matcher)``: matches an iterable if every element matches `matcher`.
```
This name feels ambiguous to me, since it could also mean the same thing as, say, `contains_exactly` (perhaps with different behaviour with repeated elements).

Perhaps the name `all_elements` would be clearer? e.g. `all_elements(equal_to(42))`.

danodonovan replied:
`contains_only` -> `all_elements`

mwilliamson commented on README.rst:
```diff
@@ -97,6 +97,14 @@ For instance, ``has_attrs(name="bob")`` is equivalent to ``has_attrs(name=equal_
+
+  .. code:: python
+
+      assert_that(result, contains_only(has_attrs(name=is_a(str)))
```
I'd suggest using a simpler example, such as `contains_only(equal_to(42))`.

mwilliamson commented on README.rst:
```diff
@@ -97,6 +97,14 @@ For instance, ``has_attrs(name="bob")`` is equivalent to ``has_attrs(name=equal_
+  .. code:: python
+
+      assert_that(result, contains_only(has_attrs(name=is_a(str)))
+      # matches all objects with a string name attribute
```
The other comments explaining what matches use concrete examples, in the optimistic hope that that's clearer than describing in the abstract. Could this be rewritten to match the style of the other examples?

mwilliamson commented on precisely/__init__.py:
```diff
@@ -2,7 +2,7 @@
 from .core_matchers import equal_to, anything, all_of, any_of, not_
 from .object_matchers import has_attr, has_attrs, instance_of
-from .iterable_matchers import contains_exactly, includes, is_sequence
+from .iterable_matchers import contains_exactly, includes, is_sequence, contains_only
```
Could we keep the imported names in alphabetical order?

mwilliamson commented on precisely/__init__.py:
```diff
@@ -12,6 +12,7 @@
     "Matcher",
     "is_matcher",
     "contains_string",
+    "contains_only",
```
The order of strings here should match the import order.

mwilliamson commented on precisely/iterable_matchers.py:
```diff
@@ -139,4 +139,30 @@ def describe(self):
+    return IsSequenceContainsOnly(matcher)
+
+
+class IsSequenceContainsOnly(Matcher):
```
I'd expect the name of the class to be `ContainsOnlyMatcher` (or whatever matches the name of the function).

mwilliamson commented on precisely/iterable_matchers.py:
```diff
@@ -139,4 +139,30 @@ def describe(self):
+        values = list(actual)
+
+        if len(values) == 0:
+            return unmatched(_empty_iterable_description)
```
By analogy with `all()`, I'd expect the empty iterable to always match.

danodonovan replied:
I've updated the code, but this does feel a little counter intuitive to me - for example

```
     assert_that([], all_elements(equal_to(42)))
```

Can we really say that every item in an empty list is equal to anything? ie If I want to check every item in a list is 42, I would be surprised if an empty list matched.

mwilliamson commented on precisely/iterable_matchers.py:
```diff
@@ -139,4 +139,30 @@ def describe(self):
+class IsSequenceContainsOnly(Matcher):
+
+    def __init__(self, matcher):
+        self._matcher = matcher
```
Would `element_matcher` be a clearer name?

mwilliamson commented on precisely/iterable_matchers.py:
```diff
@@ -139,4 +139,30 @@ def describe(self):
+        return matched()
+
+    def describe(self):
+        return "iterable only containing: {0}".format(self._matcher.describe())
```
Along the same lines as the name of the matcher, "only containing" feels ambiguous to me.

danodonovan commented:
I hope I've addressed your suggestions, always matching an empty list as True seems counter-intuitive, but I admit I don't understand all the nuances of this library. 

I use this `all_elements` feature in several repositories - having it available from master / Pypi would simplify some build processes for me!

mwilliamson commented:
I think it's the most consistent with other systems e.g. "for all" in
predicate logic, the behaviour of "all" in Python. For instance, in
Python:

>>> all([])
True

On Wed, 21 Nov 2018 05:05:30 -0800
danodonovan <notifications@github.com> wrote:

> danodonovan commented on this pull request.
> 
> 
> 
> > @@ -139,4 +139,30 @@ def describe(self):  
>              ))
>  
>  
> +def contains_only(matcher):
> +    return IsSequenceContainsOnly(matcher)
> +
> +
> +class IsSequenceContainsOnly(Matcher):
> +
> +    def __init__(self, matcher):
> +        self._matcher = matcher
> +
> +    def match(self, actual):
> +        values = list(actual)
> +
> +        if len(values) == 0:
> +            return unmatched(_empty_iterable_description)
> 
> I've updated the code, but this does feel a little counter intuitive
> to me - for example
> 
> ```
>      assert_that([], all_elements(equal_to(42)))
> ```
> 
> Can we really say that every item in an empty list is equal to
> anything? ie If I want to check every item in a list is 42, I would
> be surprised if an empty list matched. 
>

mwilliamson commented:
Looks good, thanks.
